src/services/review/RepoCheckerService.ts

```diff
@@ -4,110 +4,358 @@ import path from "path"
-                // Amplify supports adding of trailing slash
-                !setOfAllMediaAndPagesPath.has(`${encodedHref}/`)
-              ) {
-                console.log(setOfAllMediaAndPagesPath)
-                this.persistLogs(repo, [
-                  new SiteCheckerError(
-                    `Broken link ${href} that links ${text} found in ${fileName} which has permalink ${filePermalink}`
-                  ),
-                ])
-              }
+        .andThen((setOfAllMediaAndPagesPath) => {
+          mdFiles.forEach(async (fileName) => {
+            const file = fs.readFileSync(path.join(repoPath, fileName), "utf8")
+
+            const filePermalink = file
+              .split("---")
+              ?.at(1)
+              ?.split("permalink: ")
+              ?.at(1)
+              ?.split("\n")
+              ?.at(0)
+
+            if (!filePermalink) {
+              return
             }
-          })
-          const imgTags = dom.window.document.querySelectorAll("img")
-          forEach(imgTags, (tag) => {
-            const src = tag.getAttribute("src")
-            if (src) {
-              const encodedSrc = this.documentPathDecoder(src)
-              if (
-                !this.isExternalLinkOrPageRef(encodedSrc) &&
-                !setOfAllMediaAndPagesPath.has(encodedSrc)
-              ) {
-                this.persistLogs(repo, [
-                  new SiteCheckerError(
-                    `Broken image ${src} found in ${fileName} which has permalink ${filePermalink}`
-                  ),
-                ])
-              }
+
+            // we are only parsing out the front matter
+            const fileContent = file.split("---")?.slice(2).join("---")
+            if (!fileContent) {
+              return
             }
+            const html = await marked.parse(fileContent)
+            const dom = new JSDOM(html)
+            const anchorTags = dom.window.document.querySelectorAll("a")
+            forEach(anchorTags, (tag) => {
+              const href = tag.getAttribute("href")
+              const text = tag.textContent || ""
+              if (href) {
+                const encodedHref = this.mediaPathDecoder(href)
+                if (
+                  !this.isExternalLinkOrPageRef(encodedHref) &&
+                  !setOfAllMediaAndPagesPath.has(encodedHref) &&
+                  // Amplify supports adding of trailing slash
+                  !setOfAllMediaAndPagesPath.has(`${encodedHref}/`)
+                ) {
+                  const error: RepoError = {
+                    type: RepoErrorTypes.BROKEN_LINK,
+                    linkToAsset: href,
+
+                    viewablePageInCms: this.getPathInCms(fileName),
+                    viewablePageInStaging: path.join(
+                      stagingLink,
+                      filePermalink
+                    ),
+                    linkedText: text,
+                  }
+                  errors.push(error)
```

Comment:
since this is a proto, ok w/ this for now but this is a code smell. the error here is a derived state and we can jsut return it directly

Reply:
hmm, no not exactly - the `error` here is a derived state of our repos.

we shouldn't need to declare it at hte top and then do a `.push` - this loses locality and makes us prone to errors (scope is within lifetime of function).

we might, instead, want to have smaller methods that just return each category of error and we do the combine ourselves:

```ts
.andThen(prevErrors => { 
  const brokenLinks = checkForBrokenLinks(file)
  return ok([...prevErrors, ...brokenLinks])
})
```